// @sweetalert2/ngx-sweetalert2/esm5/lib/swal-portal-targets.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/swal-portal-targets.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * @record
 */
export function SwalPortalTarget() { }
if (false) {
    /** @type {?|undefined} */
    SwalPortalTarget.prototype.options;
    /**
     * @param {?} swal
     * @return {?}
     */
    SwalPortalTarget.prototype.element = function (swal) { };
}
/**
 * Represents an object of targets for <swal> portals (use with *swalPortal directive).
 * We must use thunks to access the Swal.* functions listed below, because they get created after the first modal is
 * shown, so this object lets us reference those functions safely and in a statically-typed manner.
 */
var SwalPortalTargets = /** @class */ (function () {
    function SwalPortalTargets() {
        /**
         * Targets the modal close button block contents.
         */
        this.closeButton = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getCloseButton(); }),
            options: { showCloseButton: true }
        };
        /**
         * Targets the modal title block contents.
         */
        this.title = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getTitle(); }),
            // Empty text that will never be shown but necessary so SweetAlert2 makes the div visible.
            options: { title: ' ' }
        };
        /**
         * Targets the modal text block contents (that is another block inside the first content block, so you can still
         * use other modal features like Swal inputs, that are situated inside that parent content block).
         */
        this.content = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getHtmlContainer(); }),
            // Empty text that will never be shown but necessary so SweetAlert2 makes the div visible.
            options: { text: ' ' }
        };
        /**
         * Targets the actions block contents, where are the confirm and cancel buttons in a normal time.
         * /!\ WARNING: using this target destroys some of the native SweetAlert2 modal's DOM, therefore, if you use this
         *     target, do not update the modal via <swal> \@Inputs while the modal is open, or you'll get an error.
         *     We could workaround that inconvenient inside this integration, but that'd be detrimental to memory and
         *     performance of everyone, for a relatively rare use case.
         */
        this.actions = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getActions(); }),
            // The button will never exist, but SweetAlert2 shows the actions block only if there is at least one button.
            options: { showConfirmButton: true }
        };
        /**
         * Targets the confirm button contents, replacing the text inside it (not the button itself)
         */
        this.confirmButton = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getConfirmButton(); }),
            options: { showConfirmButton: true }
        };
        /**
         * Targets the cancel button contents, replacing the text inside it (not the button itself)
         */
        this.cancelButton = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getCancelButton(); }),
            options: { showCancelButton: true }
        };
        /**
         * Targets the modal footer contents.
         */
        this.footer = {
            element: (/**
             * @param {?} swal
             * @return {?}
             */
            function (swal) { return swal.getFooter(); }),
            // Empty text that will never be shown but necessary so SweetAlert2 makes the div visible.
            options: { footer: ' ' }
        };
    }
    SwalPortalTargets.decorators = [
        { type: Injectable, args: [{ providedIn: 'root' },] }
    ];
    /** @nocollapse */ SwalPortalTargets.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function SwalPortalTargets_Factory() { return new SwalPortalTargets(); }, token: SwalPortalTargets, providedIn: "root" });
    return SwalPortalTargets;
}());
export { SwalPortalTargets };
if (false) {
    /**
     * Targets the modal close button block contents.
     * @type {?}
     */
    SwalPortalTargets.prototype.closeButton;
    /**
     * Targets the modal title block contents.
     * @type {?}
     */
    SwalPortalTargets.prototype.title;
    /**
     * Targets the modal text block contents (that is another block inside the first content block, so you can still
     * use other modal features like Swal inputs, that are situated inside that parent content block).
     * @type {?}
     */
    SwalPortalTargets.prototype.content;
    /**
     * Targets the actions block contents, where are the confirm and cancel buttons in a normal time.
     * /!\ WARNING: using this target destroys some of the native SweetAlert2 modal's DOM, therefore, if you use this
     *     target, do not update the modal via <swal> \@Inputs while the modal is open, or you'll get an error.
     *     We could workaround that inconvenient inside this integration, but that'd be detrimental to memory and
     *     performance of everyone, for a relatively rare use case.
     * @type {?}
     */
    SwalPortalTargets.prototype.actions;
    /**
     * Targets the confirm button contents, replacing the text inside it (not the button itself)
     * @type {?}
     */
    SwalPortalTargets.prototype.confirmButton;
    /**
     * Targets the cancel button contents, replacing the text inside it (not the button itself)
     * @type {?}
     */
    SwalPortalTargets.prototype.cancelButton;
    /**
     * Targets the modal footer contents.
     * @type {?}
     */
    SwalPortalTargets.prototype.footer;
}
