// @sweetalert2/ngx-sweetalert2/esm5/lib/sweetalert2-loader.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/sweetalert2-loader.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable } from '@angular/core';
import { swalProviderToken } from './di';
var SweetAlert2LoaderService = /** @class */ (function () {
    // Using any because Angular metadata generator does not understand a pure TS type here
    function SweetAlert2LoaderService(swalProvider) {
        this.swalProvider = swalProvider;
    }
    Object.defineProperty(SweetAlert2LoaderService.prototype, "swal", {
        get: /**
         * @return {?}
         */
        function () {
            if (!this.swalPromiseCache) {
                this.preloadSweetAlertLibrary();
            }
            return (/** @type {?} */ (this.swalPromiseCache));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SweetAlert2LoaderService.prototype.preloadSweetAlertLibrary = /**
     * @return {?}
     */
    function () {
        if (this.swalPromiseCache)
            return;
        /** @type {?} */
        var libPromise = isLoader(this.swalProvider)
            ? this.swalProvider()
            : Promise.resolve(this.swalProvider);
        this.swalPromiseCache = libPromise.then((/**
         * @param {?} value
         * @return {?}
         */
        function (value) { return isDefaultExport(value) ? value : value.default; }));
        /**
         * @param {?} value
         * @return {?}
         */
        function isLoader(value) {
            return typeof value === 'function' && ((/** @type {?} */ (value))).version === undefined;
        }
        /**
         * @param {?} value
         * @return {?}
         */
        function isDefaultExport(value) {
            return typeof value === 'function';
        }
    };
    SweetAlert2LoaderService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SweetAlert2LoaderService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [swalProviderToken,] }] }
    ]; };
    return SweetAlert2LoaderService;
}());
export { SweetAlert2LoaderService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    SweetAlert2LoaderService.prototype.swalProvider;
    /**
     * @type {?}
     * @private
     */
    SweetAlert2LoaderService.prototype.swalPromiseCache;
}
